### PR TITLE
[Metric] Add common metrics

### LINKED ISSUE #2743: [Metric]Add metrics
## Background
There is no metrics in mars and it's useful and necessary to add a simple and convenient metric framework.

## Design
`Counter`, `Gauge`, `Meter` and `Histogram` are the most commonly used metrics. 
* `Counter` is a cumulative type of data which represents a monotonically increasing number
* `Gauge` is a single numerical value
* `Meter` is the rate at which a set of events occur. we can use it as qps or tps
* `Histogram` records the average value of a window data

And we can use these four types of metrics as follows:
```
# Counter can be declared as `Counter(name, description: str = "", tag_keys: Optional[Tuple[str]] = None)` 
# and the others  are similar.

c1 = Counter('counter1', 'A counter')
c1.record(1)

c2 = Counter('counter2', 'A counter', ('service', 'tenant'))
c2.record(1, {'service': 'mars', 'tenant': 'test'})

g1 = Gauge('gauge1')
g1.record(1)

m1 = Meter('meter1')
m1.record(1)

h1 = Histogram('histogram1')
h1.record(1)
```

We'll implement 3 types backends of metrics including console which just log the value and can be used to debug, ray and prometheus.

And there will be 4 PRs:
- [x] First pr adds metrics framework and console backend(#2742)
- [x] Second pr adds ray implementation(#2749)
- [x] Third pr adds prometheus implementation(#2752)
- [ ] The last pr adds common metrics for mars like tileable graph building time and subtask execution time(#2760)

And we propose a naming convention for metrics as follows:
`[namespace].[component]_name[_units]`
namespace could be mars, component could be supervisor or worker and can be omitted.
For example, we can naming a `mars.subtask_execution_time_secs` to present the subtask execution time.

### PR DESCRIPTION
**Pull request by zhongchun**
## What do these changes do?
This pr adds custom defined metrics to record graph building, subtask execution time.
<div class="lake-content" typography="classic">

Metric name | Tag | Meaning | Source
-- | -- | -- | --
mars.tileable_graph_gen_time_secs | addresssession_id | Time consuming in seconds to generate a tileable graph | drvier
mars.chunk_graph_gen_time_secs | session_idtask_id | Time consuming in seconds to generate a chunk graph | supervisor
mars.subtask_graph_gen_time_secs | session_idtask_idstage_id | Time consuming in seconds to generate a subtask graph | supervisor
mars.subtask_execution_time_secs | session_idsubtask_id | Time consuming in seconds to execute a subtask | worker subpool
mars.stage_execution_time_secs | session_idtask_idstage_id | Time consuming in seconds to execute a stage | supervisor
mars.task_execution_time_secs | session_idtask_id | Time consuming in seconds to execute a task | supervisor

</div>

And we propose a naming convention for metrics as follows:
```
[namespace].[component]_name[_units]
```
`namespace` could be mars, `component` could be supervisor or worker and can be omitted.
For example, we can naming a `mars.subtask_execution_time_secs` to present the subtask execution time.

## Related issue number
Closes #2743 

## Check code requirements

- [ ] tests added / passed (if needed)
- [ ] Ensure all linting tests pass, see [here](https://docs.pymars.org/en/latest/development/contributing.html#check-code-styles) for how to run them

### CODE REVIEW

**Comment by qinxuye** [comment]
Is it possible to add more metrics, like subtask prepare data time(including fetch data from other workers), subtask execution time(real execution time), subtask store time, etc?

**Comment by zhongchun** [comment]
> Is it possible to add more metrics, like subtask prepare data time(including fetch data from other workers), subtask execution time(real execution time), subtask store time, etc?
> 
> Do you think shall us add them in this PR or maybe sometimes later?

Yes, we'll add more metrics like you say. This is the first pr adding custom metrics and bringing metric configs. And these metrics are basic, commonly used. It's difficult to add all metrics at once. Maybe we can add more metrics in later PRs.

**Comment by qinxuye** [comment]
> > Is it possible to add more metrics, like subtask prepare data time(including fetch data from other workers), subtask execution time(real execution time), subtask store time, etc?
> > Do you think shall us add them in this PR or maybe sometimes later?
> 
> Yes, we'll add more metrics like you say. This is the first pr adding custom metrics and bringing metric configs. And these metrics are basic, commonly used. It's difficult to add all metrics at once. Maybe we can add more metrics in later PRs.

Reasonable to me.

**Review by qinxuye — DISMISSED** [review]
LGTM

File: mars/deploy/oscar/base_config.yml [hunk 1]
```diff
@@ -36,3 +36,6 @@ scheduling:
     max_workers: 100
     scheduler_backlog_timeout: 60
     worker_idle_timeout: 120
+global_config:
```
Comment (wjsi) [inline mars/deploy/oscar/base_config.yml:39]: For me, there is no need to add a separate section called `global_config` as the name is somehow confusing. Just put `metrics` as a root sector will look better for me.

File: mars/deploy/oscar/rayconfig.yml [hunk 2]
```diff
@@ -14,3 +14,6 @@ scheduling:
     enabled: false
     scheduler_backlog_timeout: 20
     worker_idle_timeout: 40
+global_config:
```
Comment (wjsi) [inline mars/deploy/oscar/rayconfig.yml:17]: Ditto.

File: mars/deploy/oscar/local.py [hunk 3]
```diff
@@ -216,6 +217,7 @@ async def _start_worker_pools(self):
                 band_to_slot,
                 modules=worker_modules,
                 subprocess_start_method=self._subprocess_start_method,
+                _global_config=self._config.get("global_config", {}),
```
Comment (wjsi) [inline mars/deploy/oscar/local.py:220]: Ditto.

File: mars/deploy/oscar/ray.py [hunk 4]
```diff
@@ -198,7 +199,10 @@ async def new_worker(self, worker_address, band_to_slot=None):
         start_time = time.time()
         band_to_slot = band_to_slot or self._band_to_slot
         worker_pool = await create_worker_actor_pool(
-            worker_address, self._band_to_slot, modules=self._worker_modules
+            worker_address,
+            self._band_to_slot,
+            modules=self._worker_modules,
+            _global_config=self._config.get("global_config", {}),
```
Comment (wjsi) [inline mars/deploy/oscar/ray.py:205]: Ditto.

File: mars/deploy/oscar/session.py [hunk 5]
```diff
@@ -34,6 +34,7 @@
 import numpy as np
 
 from ... import oscar as mo
+from ..._utils import Timer
```
Comment (wjsi) [inline mars/deploy/oscar/session.py:37]: `_utils` is not intended for use outside root package. Use `utils` directly.

File: mars/oscar/backends/config.py [hunk 6]
```diff
@@ -26,6 +26,8 @@ def __init__(self, conf: Dict = None):
             self._conf["pools"] = dict()
         if "mapping" not in self._conf:
             self._conf["mapping"] = dict()
+        if "global_config" not in self._conf:
```
Comment (wjsi) [inline mars/oscar/backends/config.py:29]: Naming issue as is referred to above.

File: mars/deploy/oscar/ray.py [hunk 7]
```diff
@@ -402,6 +406,8 @@ def __init__(
         self.web_address = None
 
     async def start(self):
+        # init metrics to guarantee metrics use in driver
+        init_metrics(self._config.get("global_config", {}))
```
Comment (wjsi) [inline mars/deploy/oscar/ray.py:410]: The function should also consider using more explicit argument lists instead of dicts.

**Review by wjsi — COMMENTED** [review]
Besides these comments, also consider unifying implementations with #2586 for further expansion. We cannot accept metrics or profiles with different implementations.

File: mars/deploy/oscar/ray.py (hunk 4 above)
Comment (zhongchun) [reply mars/deploy/oscar/ray.py:205]: I'll fix it.

File: mars/oscar/backends/config.py (hunk 6 above)
Comment (zhongchun) [reply mars/oscar/backends/config.py:29]: Thanks, i'll fix it.

File: mars/deploy/oscar/ray.py (hunk 7 above)
Comment (zhongchun) [reply mars/deploy/oscar/ray.py:410]: Good suggestion. Thanks.

**Comment by zhongchun** [comment]
> Besides these comments, also consider unifying implementations with #2586 for further expansion. We cannot accept metrics or profiles with different implementations.

Yes. Agree with you. Profiles and metrics are used in different scenarios, but we will keep their data unique and same.

**Comment by qinxuye** [comment]
Metrics and enable_profiling look pretty overlapped, can enable_profiling be merged into metrics some day in the near future, after all, console metric is enabled? @fyrestone , what's your opinion?

And, how about rename `metric` directory to `metrics` which looks more consistent? @zhongchun

**Comment by zhongchun** [comment]
> Metrics and enable_profiling look pretty overlapped, can enable_profiling be merged into metrics some day in the near future, after all, console metric is enabled? @fyrestone , what's your opinion?
> 
> And, how about rename `metric` directory to `metrics` which looks more consistent? @zhongchun

Year,`metrics` is fine. 

Metrics and profile may use same data, but not they are different. Perhaps we should enable them separately.

**Comment by fyrestone** [comment]
> consistent

Profiling only gives general statistics info, the result of profiling should be concise. While metrics logs all the infos to metrics systems. I can remove the profiling code if metrics is enough for profiling.

File: mars/services/subtask/core.py [hunk 8]
```diff
@@ -106,13 +106,28 @@ class SubtaskResult(Serializable):
...
             self.bands = sorted(set(bands + result.bands))
+            self._update_result(result)
         return self
 
+    def _update_result(self, result: Optional["SubtaskResult"]):
+        """Update result to save the latest successful execution result."""
+        if (
+            self.status != SubtaskStatus.succeeded
+            and result.status == SubtaskStatus.succeeded
+        ):
```
Comment (wjsi) [inline mars/services/subtask/core.py:125]: `execution_end_time` shall be set even if `result.ststus` is errored or cancelled.

File: mars/services/subtask/core.py [hunk 9]
```diff
@@ -106,13 +106,28 @@ class SubtaskResult(Serializable):
...
+    execution_end_time: float = Float64Field("execution_end_time")
 
     def merge_bands(self, result: Optional["SubtaskResult"]):
         if result and result.bands:
             bands = self.bands or []
             self.bands = sorted(set(bands + result.bands))
+            self._update_result(result)
         return self
 
+    def _update_result(self, result: Optional["SubtaskResult"]):
```
[the remark this answers is not in the document]
Comment (zhongchun) [reply mars/services/subtask/core.py:120]: OK, thanks.

File: mars/services/subtask/core.py (hunk 8 above)
Comment (zhongchun) [reply mars/services/subtask/core.py:125]: Agree with you. I'll fix it.

Comment (zhongchun) [reply mars/services/subtask/core.py:125]: I just remove `_update_result` method and put them in the `if` block for simplicity and clarity.

File: mars/services/subtask/core.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (zhongchun) [reply mars/services/subtask/core.py:120]: I just remove `_update_result` method and put them in the `if` block for simplicity and clarity.

**Comment by wjsi** [comment]
Please do not submit empty commits. This will retry all tasks and block other runners. We can rerun failed tasks by hand, or you may ask [a project member](https://github.com/mars-project/mars/people) for help.

**Comment by zhongchun** [comment]
> Please do not submit empty commits. This will retry all tasks and block other runners. We can rerun failed tasks by hand, or you may ask [a project member](https://github.com/mars-project/mars/people) for help.

Thank you for reminding.

**Comment by qinxuye** [comment]
Can this PR close #2743 ? I suggest so, this is the last step to introduce metrics, further issues and PRs can be submitted later.

**Comment by zhongchun** [comment]
> 

Yes.

**Comment by zhongchun** [comment]
> Can this PR close #2743 ? I suggest so, this is the last step to introduce metrics, further issues and PRs can be submitted later.

Yes.

**Comment by qinxuye** [comment]
Seems some tests cannot pass, a few timeouts, can u see what's happened there?

**Comment by zhongchun** [comment]
> Seems some tests cannot pass, a few timeouts, can u see what's happened there?

Thanks for your remanding. I'll fix them.

**Review by wjsi — APPROVED** [review]
LGTM

**Review by hekaisheng — APPROVED** [review]
LGTM

File: mars/deploy/oscar/tests/test_ray.py [hunk 10]
```diff
@@ -635,3 +635,15 @@ async def test_ownership_when_scale_in(ray_large_cluster):
         assert (
             groupby_sum_df.to_pandas().to_dict() == pd_df.groupby("a").sum().to_dict()
         )
+
+
+@require_ray
+@pytest.mark.asyncio
+def test_init_metrics_on_ray(ray_large_cluster, create_cluster):
```
Comment (chaokunyang) [inline mars/deploy/oscar/tests/test_ray.py:642]: use `ray_start_regular` instead

File: mars/services/subtask/worker/processor.py [hunk 11]
```diff
@@ -489,6 +501,20 @@ async def run(self):
                 await self._unpin_data(input_keys)
 
         await self.done()
+        if self.result.status == SubtaskStatus.succeeded:
+            cost_time_secs = (
+                self.result.execution_end_time - self.result.execution_start_time
+            )
+            logger.info(
+                "Time consuming to execute a subtask is %ss with session_id %s, subtask_id %s",
```
Comment (chaokunyang) [inline mars/services/subtask/worker/processor.py:509]:
```suggestion
                "Executing subtask %s with session_id %s took %s",
```

**Review by chaokunyang — APPROVED** [review]
LGTM

**Comment by hekaisheng** [comment]
> LGTM

Why approve while left these comments, seems "request changes" is more appropriate.

File: mars/deploy/oscar/tests/test_ray.py (hunk 10 above)
Comment (zhongchun) [reply mars/deploy/oscar/tests/test_ray.py:642]: Ok. I'll fix it in later pr.

File: mars/services/subtask/worker/processor.py (hunk 11 above)
Comment (zhongchun) [reply mars/services/subtask/worker/processor.py:509]: Good suggestion. I'll change it in later pr.

**Pull request merged**